bin/varnishd/cache/cache_deliver_proc.c

```diff
@@ -83,19 +109,20 @@ VDP_push(struct req *req, const struct vdp *vdp, void *priv, int bottom)
 	vdc = req->vdc;
 	AN(vdp);
 	AN(vdp->name);
-	AN(vdp->func);
+	AN(vdp->bytes);
 
-	if (vdc->retval)
-		return (vdc->retval);
+	if (vdc->status != VDP_OK)
+		return (vdc->status);
 
 	if (DO_DEBUG(DBG_PROCESSORS))
 		VSLb(req->vsl, SLT_Debug, "VDP_push(%s)", vdp->name);
 
 	vdpe = WS_Alloc(req->ws, sizeof *vdpe);
 	if (vdpe == NULL) {
-		AZ(vdc->retval);
-		vdc->retval = -1;
-		return (vdc->retval);
+		VSLb(req->vsl, SLT_Error,
```

Comment:
why pull in the error reporting? We got more error cases in the caller, so keeping the reporting in the caller would appear to be the natural choice?

Reply:
My thought on the error reporting was to allow the one noticing the error to do the reporting, as it is best equiped to do a specific error text rather than just a generic "VDP initialization error". This particular line is the VDP_push reporting that it failed to set up the VDP stack because of WS exhaustion.

In general, a VDP that wants to fail the delivery during its initialization should log appropriately before returning error. But since we can't trust a random VMOD to adhere to that rule, we also log a generic  "VDP initialization error" if things went south. This will result in multiple SLT_Error log lines instead of a single just to be sure. If we don't like this, we maybe should adopt a something like in VFP_Error instead, where we record if an error message has already been printed, and not do it twice.

Reply:
Yeah something like VDP_Error sounds like a sensible thing to do. This way we can achieve your sensible goal to ensure there always is an error emitted while still keeping it concise inside the VDP.